File: python/0498_diagonal_traverse.py
"""
Problem: 498 - Diagonal Traverse  
Description: Given an m x n matrix mat, return an array of all the elements of the matrix in diagonal order. The traversal should alternate between upward and downward diagonals.  
Date: 2025-08-25
"""

class Solution:
    def findDiagonalOrder(self, mat: List[List[int]]) -> List[int]:
        m,n = len(mat), len(mat[0])

        diagonals = defaultdict(list)
        for i in range(m):
            for j in range(n):
                diagonals[i + j].append(mat[i][j])

        result = []
        for k in range(m + n - 1):
            if k % 2 == 0:
                result.extend(reversed(diagonals[k]))
            else:
                result.extend(diagonals[k])
        
        return result

File: java/0498_diagonal_traverse.java
"""
Problem: 498 - Diagonal Traverse  
Description: Given an m x n matrix mat, return an array of all the elements of the matrix in diagonal order. The traversal should alternate between upward and downward diagonals.  
Date: 2025-08-25
"""

class Solution {
    public int[] findDiagonalOrder(int[][] mat) {
        Map<Integer, List<Integer>> diagonals = new HashMap<>();
        int m = mat.length, n = mat[0].length;

        for (int i = 0; i < m; i++)
            for (int j = 0; j < n; j++)
                diagonals.computeIfAbsent(i + j, x -> new ArrayList<>()).add(mat[i][j]);

        List<Integer> result = new ArrayList<>();
        for (int k = 0; k < m + n - 1; k++) {
            List<Integer> diag = diagonals.get(k);
            if (k % 2 == 0)
                Collections.reverse(diag);
            result.addAll(diag);
        }

        return result.stream().mapToInt(i -> i).toArray();
    }
}
